projects/RabbitMQ.Client/util/PipeReaderExtensions.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System;
+using System.Buffers;
+using System.IO;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+
+namespace RabbitMQ.Client
+{
+    internal static class PipeReaderExtensions
```

Comment:
whats the reason for all these extensions? why is the copy on read needed? looks like a waste to me

Reply:
The short version, Pipelines reuse internal buffers, so you have to copy the content you want to keep from them befire calling AdvanceTo, since that'll return the buffer for reuse within the Pipelines later.